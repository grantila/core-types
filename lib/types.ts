
export interface LineColumn
{
	line: number;
	column: number;
	offset: number;
}

export interface LocationWithLineColumn {
    start?: LineColumn;
    end?: LineColumn;
}

export interface LocationOffset
{
	start?: number;
	end?: number;
}

export type Location =
	| LocationWithLineColumn
	| LocationOffset;

export interface CoreTypeAnnotations
{
	name?: string;
	title?: string;
	description?: string;
	examples?: string | Array< string >;
	default?: string;
	see?: string | Array< string >;
	comment?: string;
	loc?: LocationOffset | LocationWithLineColumn;
}

export interface AndType extends CoreTypeAnnotations
{
	type: 'and';
	and: NodeType[ ];
}

export interface OrType extends CoreTypeAnnotations
{
	type: 'or';
	or: NodeType[ ];
}

export interface TypeMap
{
	and: void;
	or: void;
	ref: string;
	any: unknown;
	null: null;
	string: string;
	number: number;
	integer: number;
	boolean: boolean;
	object: object;
	array: Array< unknown >;
	tuple: Array< unknown >;
}

export type Types = keyof TypeMap;

export interface Const< T >
{
	const?: T;
}

export interface Enum< T >
{
	enum?: Array< T >;
}

export type GenericTypeInfo< T > =
	& Const< T >
	& Enum< T >;

export interface NodePrimitiveCoreType<
	Type extends 'any' | 'null' | 'string' | 'number' | 'integer' | 'boolean'
>
{
	type: Type;
}

export type NodePrimitiveType<
	Type extends 'any' | 'null' | 'string' | 'number' | 'integer' | 'boolean'
> =
	& NodePrimitiveCoreType< Type >
	& GenericTypeInfo< TypeMap[ Type ] >;

export type AnyType = NodePrimitiveType< 'any' > & CoreTypeAnnotations;
export type NullType = NodePrimitiveCoreType< 'null' > & CoreTypeAnnotations;
export type StringType = NodePrimitiveType< 'string' > & CoreTypeAnnotations;
export type NumberType = NodePrimitiveType< 'number' > & CoreTypeAnnotations;
export type IntegerType = NodePrimitiveType< 'integer' > & CoreTypeAnnotations;
export type BooleanType = NodePrimitiveType< 'boolean' > & CoreTypeAnnotations;

export type PrimitiveType =
	| NullType
	| StringType
	| NumberType
	| IntegerType
	| BooleanType;

export interface NodeRefCoreType
{
	type: 'ref';
	ref: string;
}

export type RefType =
	& NodeRefCoreType
	& GenericTypeInfo< unknown >
	& CoreTypeAnnotations;

export interface ObjectProperty
{
	required: boolean;
	node: NodeType;
}

export interface NodeObjectCoreType
{
	type: 'object';
	properties: {
		[ name: string ]: ObjectProperty;
	};
	additionalProperties: boolean | NodeType;
}

export type ObjectType =
	& NodeObjectCoreType
	& GenericTypeInfo< object >
	& CoreTypeAnnotations;

export interface NodeArrayCoreType
{
	type: 'array';
	elementType: NodeType;
}

export type ArrayType< T = unknown > =
	& NodeArrayCoreType
	& GenericTypeInfo< Array< T > >
	& CoreTypeAnnotations;

export interface NodeTupleCoreType
{
	type: 'tuple';
	elementTypes: Array< NodeType >;
	minItems: number;
	additionalItems: boolean | NodeType;
}

export type TupleType< T extends unknown[ ] = unknown[ ] > =
	& NodeTupleCoreType
	& GenericTypeInfo< T >
	& CoreTypeAnnotations;

export interface NodeTypeMap
{
	and: AndType;
	or: OrType;
	ref: RefType;
	any: AnyType;
	null: NullType;
	string: StringType;
	number: NumberType;
	integer: IntegerType;
	boolean: BooleanType;
	object: ObjectType;
	array: ArrayType;
	tuple: TupleType;
}

export type NodeType = NodeTypeMap[ keyof NodeTypeMap ];

export type NamedType< T extends NodeType = NodeType > = T & { name: string; };

export type NodeWithConstEnum =
	| AnyType
	| StringType
	| NumberType
	| IntegerType
	| BooleanType
	| ObjectType
	| ArrayType
	| TupleType
	| RefType;

export type NodePath = Array< string | number >;

export interface NodeDocument<
	Version extends number = 1,
	T extends NodeType = NodeType
>
{
	version: Version;
	types: Array< NamedType< T > >;
}

export interface ConversionResult< T = string >
{
	data: T;
	convertedTypes: Array< string >;
	notConvertedTypes: Array< string >;
}
