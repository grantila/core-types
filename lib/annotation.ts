import { mergeLocations } from "./location";
import type { CoreTypeAnnotations, NodeType } from "./types"
import { ensureArray } from "./util"


export function mergeAnnotations( nodes: Array< CoreTypeAnnotations > )
: CoreTypeAnnotations
{
	const nonEmpty = < T >( t: T ): t is NonNullable< T > => !!t;
	const join = < T >( t: Array< T > ) =>
		t.filter( nonEmpty ).join( "\n" ).trim( );

	const name = nodes.find( n => n.name )?.name;
	const title = join( nodes.map( n => n.title ) );
	const description = join( nodes.map( n => n.description ) );
	const examples =
		( [ ] as Array< string > ).concat(
			...nodes.map( n => ensureArray( n.examples ) )
		)
		.filter( nonEmpty );
	const _default = join( nodes.map( n => n.default ) );
	const see =
		( [ ] as Array< string > ).concat(
			...nodes.map( n => ensureArray( n.see ) )
		)
		.filter( nonEmpty );
	const comment = join( nodes.map( n => n.comment ) );
	const loc = mergeLocations( nodes.map( n => n.loc ) );

	return {
		...( name ? { name } : { } ),
		...( title ? { title } : { } ),
		...( description ? { description } : { } ),
		...(
			examples.length > 0
			? { examples: arrayOrSingle( examples ) }
			: { }
		),
		...( _default ? { default: _default } : { } ),
		...(
			see.length > 0
			? { see: arrayOrSingle( see ) }
			: { }
		),
		...( comment ? { comment } : { } ),
		...( loc ? { loc } : { } ),
	};
}

export function extractAnnotations( node: CoreTypeAnnotations )
: CoreTypeAnnotations
{
	const {
		title,
		description,
		examples,
		default:
		_default,
		comment,
		see,
	} = node;

	return {
		...( title ? { title } : { } ),
		...( description ? { description } : { } ),
		...( examples ? { examples } : { } ),
		...( _default ? { default: _default } : { } ),
		...( comment ? { comment } : { } ),
		...( see ? { see } : { } ),
	};
}

export interface StringifyAnnotationsOptions
{
	/**
	 * Include the comment part of the annotation
	 */
	includeComment?: boolean;

	/**
	 * Format whitespace so that the comment can be either wrapped within
	 * a \/* and *\/ boundary or prefixed with //
	 */
	formatWhitespace?: boolean;
}

function wrapWhitespace( text: string ): string
{
	if ( !text.includes( "\n" ) )
		return text.startsWith( " " ) ? text : ` ${text}`;

	return [
		"*",
		text.split( "\n" ).map( line => ` * ${line}` ).join( "\n" ),
		" "
	].join( "\n" );
}

function makeSafeComment( text: string ): string
{
	return text.replace( /\*\//g, '*\\/' );
}

export function stringifyAnnotations(
	node: CoreTypeAnnotations,
	{
		includeComment = false,
		formatWhitespace = false,
	}: StringifyAnnotationsOptions = { }
)
: string
{
	const { title, description, examples, default: _default, comment, see } = node;
	const fullComment = makeSafeComment(
		[
			title,
			description,
			...( examples == undefined ? [ ] : [
				formatExamples( ensureArray( examples ) )
			] ),
			...( _default === undefined ? [ ] : [
				formatDefault( _default )
			] ),
			...( see == undefined ? [ ] : [
				formatSee( ensureArray( see ) )
			] ),
			...( includeComment ? [ comment ] : [ ] ),
		]
		.filter( v => v )
		.join( "\n\n" )
		.trim( )
	);

	return formatWhitespace && fullComment
		? wrapWhitespace( fullComment )
		: fullComment;
}

export function stripAnnotations< T extends NodeType >(
	node: T,
	recursive = true
)
: T
{
	const {
		comment,
		description,
		default: _default,
		examples,
		see,
		title,
		...rest
	} = node;

	const filteredNode = rest as NodeType & T;

	if ( recursive )
	{
		if ( filteredNode.type === 'and' )
			return {
				...filteredNode,
				and: filteredNode.and.map( n => stripAnnotations( n, true ) ),
			};
		else if ( filteredNode.type === 'or' )
			return {
				...filteredNode,
				or: filteredNode.or.map( n => stripAnnotations( n, true ) ),
			};
		else if ( filteredNode.type === 'array' )
			return {
				...filteredNode,
				elementType:
					stripAnnotations( filteredNode.elementType, true ),
			};
		else if ( filteredNode.type === 'tuple' )
			return {
				...filteredNode,
				elementTypes: filteredNode.elementTypes.map( n =>
					stripAnnotations( n, true )
				),
				additionalItems:
					typeof filteredNode.additionalItems === 'object'
					? stripAnnotations( filteredNode.additionalItems, true )
					: filteredNode.additionalItems,
			};
		else if ( filteredNode.type === 'object' )
			return {
				...filteredNode,
				properties: Object.fromEntries(
					Object.keys( filteredNode.properties ).map( key =>
						[
							key,
							{
								...filteredNode.properties[ key ],
								node: stripAnnotations(
									filteredNode.properties[ key ].node,
									true
								),
							}
						]
					)
				),
				additionalProperties:
					typeof filteredNode.additionalProperties === 'object'
					? stripAnnotations(
						filteredNode.additionalProperties,
						true
					)
					: filteredNode.additionalProperties,
			};
	}

	return filteredNode;
}

function arrayOrSingle< T >( arr: Array< T > ): T | Array< T >
{
	if ( arr.length === 1 )
		return arr[ 0 ];
	return arr;
}

function formatExamples( examples: Array< string > ): string
{
	const lines =
		examples.map( example =>
			"@example\n" + indent( stringify( example ).split( "\n" ), 4 )
		)
		.join( "\n" );

	return lines.trim( );
}

function formatDefault( _default: string ): string
{
	const lines = [
		"@default",
		indent( stringify( _default ).split( "\n" ), 4 )
	]
	.join( "\n" );

	return lines.trim( );
}

function formatSee( see: Array< string > ): string
{
	const lines =
		see.map( see => "@see " + stringify( see ) )
		.join( "\n" );

	return lines.trim( );
}

function stringify( value: any )
{
	return typeof value === "string"
		? value
		: JSON.stringify( value, null, 2 );
}

function indent( lines: Array< string >, indent: number, bullet = false )
{
	return lines
		.map( ( line, index ) =>
		{
			const prefix =
				index === 0 && bullet
				? ( ' '.repeat( indent - 2 ) + "* " )
				: ' '.repeat( indent );
			return prefix + line;
		} )
		.join( "\n" );
}
