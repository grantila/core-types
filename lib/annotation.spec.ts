import { stringifyAnnotations } from "./annotation"
import { CoreTypeAnnotations } from "./types";

describe( "annotation", ( ) =>
{
	describe( "stringifyAnnotations", ( ) =>
	{
		const getAnnotations = ( multi: boolean ): CoreTypeAnnotations => ( {
			name: "Name",
			title: "Title",
			description: "The description\non multiple lines...",
			examples: multi ? [ "Ex1", "Ex2 is here" ] : 'The example',
			default: "Joe",
			see: multi ? [ "This thing", "and this too" ] : 'Interesting',
			comment: "Should be secret by default",
		} );

		it( "should stringify everything corrently with arrays", ( ) =>
		{
			const text = stringifyAnnotations( getAnnotations( true ) );
			expect( text ).toMatchSnapshot( );
		} );

		it( "should stringify everything corrently with singles", ( ) =>
		{
			const text = stringifyAnnotations( getAnnotations( false ) );
			expect( text ).toMatchSnapshot( );
		} );
	} );
} );
