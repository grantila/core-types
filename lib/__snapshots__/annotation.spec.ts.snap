// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`annotation stringifyAnnotations should stringify everything corrently with arrays 1`] = `
"Title

The description
on multiple lines...

@example
    Ex1
@example
    Ex2 is here

@default
    Joe

@see This thing
@see and this too"
`;

exports[`annotation stringifyAnnotations should stringify everything corrently with singles 1`] = `
"Title

The description
on multiple lines...

@example
    The example

@default
    Joe

@see Interesting"
`;
